File: extension/core/src/main/java/io/holunda/camunda/bpm/data/writer/VariableMapWriter.java
package io.holunda.camunda.bpm.data.writer;

import io.holunda.camunda.bpm.data.factory.VariableFactory;
import org.camunda.bpm.engine.variable.VariableMap;
import org.jetbrains.annotations.NotNull;

import java.util.Objects;

/**
 * Variable map builder allowing for fluent variable setting.
 */
public class VariableMapWriter implements GlobalVariableWriter<VariableMapWriter> {

  private final VariableMap variables;

  /**
   * Creates a builder with provided variable map.
   *
   * @param variables variables to work on.
   */
  public VariableMapWriter(VariableMap variables) {
    this.variables = variables;
  }

  @Override
  @NotNull
  public <T> VariableMapWriter set(VariableFactory<T> factory, T value) {
    return this.set(factory, value, false);
  }

  @Override
  @NotNull
  public <T> VariableMapWriter set(VariableFactory<T> factory, T value, boolean isTransient) {
    factory.on(this.variables).set(value, isTransient);
    return this;
  }

  @Override
  @NotNull
  public <T> VariableMapWriter remove(VariableFactory<T> factory) {
    factory.on(this.variables).remove();
    return this;
  }

  @Override
  @NotNull
  public VariableMap variables() {
    return variables;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (o == null || getClass() != o.getClass()) return false;

    VariableMapWriter that = (VariableMapWriter) o;
    return Objects.equals(variables, that.variables);
  }

  @Override
  public int hashCode() {
    return variables != null ? variables.hashCode() : 0;
  }
}

File: extension/core/src/main/kotlin/io/holunda/camunda/bpm/data/writer/VariableMapWriter.kt
package io.holunda.camunda.bpm.data.writer

import io.holunda.camunda.bpm.data.factory.VariableFactory
import org.camunda.bpm.engine.variable.VariableMap
import java.util.function.Function

/**
 * Variable map builder allowing for fluent variable setting.
 * @param variables variables to work on.
 */
class VariableMapWriter(private val variables: VariableMap) : GlobalVariableWriter<VariableMapWriter> {
  override fun <T> set(variableFactory: VariableFactory<T>, value: T): VariableMapWriter {
    return this.set(variableFactory, value, false)
  }

  override fun <T> set(variableFactory: VariableFactory<T>, value: T, isTransient: Boolean): VariableMapWriter {
    variableFactory.on(variables)[value] = isTransient
    return this
  }

  override fun <T> remove(variableFactory: VariableFactory<T>): VariableMapWriter {
    variableFactory.on(variables).remove()
    return this
  }

  override fun <T> update(variableFactory: VariableFactory<T>, valueProcessor: Function<T, T>): VariableMapWriter {
    variableFactory.on(variables).update(valueProcessor)
    return this
  }

  override fun <T> update(
    variableFactory: VariableFactory<T>,
    valueProcessor: Function<T, T>,
    isTransient: Boolean
  ): VariableMapWriter {
    variableFactory.on(variables).update(valueProcessor, isTransient)
    return this
  }

  override fun variables(): VariableMap {
    return variables
  }

  override fun equals(other: Any?): Boolean {
    if (this === other) return true
    if (other == null || javaClass != other.javaClass) return false
    val that = other as VariableMapWriter
    return variables == that.variables
  }

  override fun hashCode(): Int {
    return variables.hashCode()
  }
}
